### ISSUE TITLE
multipart/form-data with uploading CSV file.

### ISSUE THREAD

**Issue by HSutrisno** [tsenart/vegeta#347]
#### Question
Hi guys I need a help for my scenario:
I have to sent http request for uploading csv file, the size of csv file is around 500mb.
I can proceed the request with curl like this:
**curl -s -o /dev/null -v -X POST -H 'document=multipart/form-data' -F 'sources=1' -F 'description=big csv file' -F 'document=@csvlog_40000.csv' http://192.167.12.119:8000/dv/validate**

but how do I achieve it by using vegeta?
I will have target.txt:
 ```
POST http://192.167.12.119:8000/dv/validate/
 Content-Type: multipart/form-data; boundary=Boundary+1234
 @/home/bootstrap/dv_test/40000_datarow/body.txt
```

and i will have body.txt:

```
--Boundary+1234
Content-Disposition: form-data; name="document"; filename="csvlog_40000.csv"
Content-Type: application/vnd.ms-excel
 --Boundary+1234
Content-Disposition: form-data; name="sources" 
  1
 --Boundary+1234
```
and my vegeta command is:
` vegeta attack -rate=1 -duration=2s -targets=target.txt > result.bin`
but it doesn't works, how should I upload csvlog_40000.csv file, is there any flag in vegeta to set the path from my local computer?therefore It can work exactly like Curl.

or vegeta doesn't support my scenario? 

body.txt is only header not the payload of data to be uploaded which is csvlog_40000.csv.

Your comment and help very appreciate so much.

Cheers.

**Comment by HarrySutrisno** [comment]
but I am not sure how he attaches the file contain, i am sure is not base on local path. 

or should i set it here:

--Boundary+0xAbCdEfGbOuNdArY
Content-Disposition: form-data; name="file"; filename="file.csv"
Content-Type: image/jpeg

contents csv file.........
...
....

--Boundary+0xAbCdEfGbOuNdArY--

**Comment by HarrySutrisno** [comment]
You are right, but if my body is like key and value:
document: content of csv to be uploaded, where should i define?

**Comment by HarrySutrisno** [comment]
@tsenart  does the body support key and value ?
something like this:

curl -X POST \
  http://192.167.12.119:8000/dv/validate/ \
  -H 'cache-control: no-cache' \
  -H 'content-type: multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW' \
  -H 'postman-token: d58e36f5-ed8b-c3fe-f05d-3d31c91b035d' \
  -F document=@checkpointlog_40000.csv \
  -F sources=1


-F is body with keys and values

**Comment by tsenart** [comment]
I don’t see why it wouldn’t. Put it in a file and try it out.

**Comment by HarrySutrisno** [comment]
I put everything in file as per below but still doesn't work:
targets.txt:
```
 POST http://192.168.11.121:8000/dv/validate/
 Host: 192.168.11.121:8000
 Cache-Control: no-cache
 Content-Type: multipart/form-data; boundary=------WebKitFormBoundaryeepPeif7gE4oPuGp
```
sample102.txt:
```
 ------WebKitFormBoundaryeepPeif7gE4oPuGp
 Content-Disposition: form-data; name="document"; filename="checkpointlog_40000.csv"
 Content-Type: application/vnd.ms-excel

Application Protocol,Bytes In,Bytes Out,External ID,Message,Transport Protocol,Start Time,Name,Event Outcome,End Time,Source Address,Source Host Name,Source Mac Address,Source Service Name,Source Translated     Address,Source User Name,Source Translated Port,Source Port,Request Url,Request Protocol,Request Cookies,Request Context,File Type,File Name,Device Action,Device Address,Device Event Category,Device Version,    Device Vendor,Device Severity,Device Host Name,Device Facility,Device Event Class ID,Device Direction,Destination Address,Destination Host Name,Destination Mac Address,Destination Port,Destination Service Name, Destination Translated Address,Destination User Name,Destination User ID,Destination Translated Port,Device Inbound Interface,Device Outbound Interface,Device Product,Device Receipt Time,Device Custom Date1,    Device Custom Date2,Device Custom IPv6 Address2,Device Custom IPv6 Address3,Device Custom Number1,Device Custom Number3,Device Custom String1,Device Custom String2,Device Custom String3,Device Custom String4,   Device Custom String5,Device Custom String6,Device Custom Number2,Manager Receipt Time
 ,,,,,TCP,Sep 15 2018 04:05:34,https_,,Sep 15 2018 04:05:34,10.204.79.187,,,,,,,63246,,,,,,,Accept,10.209.0.18,,Check Point,Check Point,Unknown,TC1CPM1.hiscox.com,,Log,Inbound,10.204.2.12,,,443,,,,,,,,VPN-1 &    FireWall-1,Sep 15 2018 04:05:34,,,,,,,,Altiris 2014 (4),,,,,,Sep 15 2018 04:05:38
 
 ------WebKitFormBoundaryeepPeif7gE4oPuGp
 Content-Disposition: form-data; name="sources"

1
 ------WebKitFormBoundaryeepPeif7gE4oPuGp--
```

```
my command:
vegeta attack -rate=1 -duration=2s -targets=targets.txt -body=sample102.txt > result.bin
```


please fix me which one is my step wrong here or missing.
it still doesn't work, I have two key and value for my body as i mentioned above:
-document
-sources

**Comment by HarrySutrisno** [comment]
I've got 500 Internal Server Error, and if you familiar with django here: https://docs.djangoproject.com/en/2.1/ref/request-response/#django.http.HttpRequest.FILES

there is attribute HttpRequest.FILES, which is will return dictionary data(http upload body) if it can manage to send/receive the  request, but right now the body is empty.

I can manage send request with 200 status code successfully by using postman and curl but not vegeta.

thanks @tsenart tsenart for your help

**Comment by tsenart** [comment]
If you use the `-body` flag then don't set a body in the targets. I am not familiar with Django unfortunately.

**Comment by tsenart** [comment]
You need to understand specifically what why your server is returning a 500. Check the logs.

**Comment by HarrySutrisno** [comment]
I have decided to use in my golang code, but I've got confused actually. What I am looking for is something like below function:
func (t *Target) Request() (*http.Request, error)

what is the criteria to use this one? may I use it as attack**(Attack(tr, rate, 1*time.Second, "")**? which is i can define the rate and duration of attack?

because in vegeta targets_test.go is only being used for normal sent request.
